### ISSUE TITLE
Mistaking macros for functions

### ISSUE THREAD

**Issue by wenda84** [microsoft/vscode-cpptools#9378]
Bug type: Language Service

**Describe the bug**

- OS and Version: win10 1909

- VS Code Version: 1.67.1 (user setup)

- C/C++ Extension Version:1.9.8

- Other extensions you installed (and if the issue persists after disabling them):

- If using SSH remote, specify OS of remote machine: displays in local environment.

- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

  mistakenly treating a macro as a function in a single project

**Steps to reproduce**

1. defined a macro for "for... loop"

   ```cpp
   #define list_for_each_safe(pos, n, head) \
   	for (pos = (head)->next, n = pos->next; pos != (head); \
   		 pos = n, n = pos->next)
   ```

   

2. use the marco in "for ... loop" in main function

   ```cpp
       list_for_each_safe(pos, tmp, &data_queue)
       {
           entry = list_entry(pos, struct data_entry, list);
           list_del(pos);
           free(entry);
       }
   ```

   

3. In VSCode outline page, treating `list_for_each_safe` as a function and will not show the main function.

   

**Expected behavior**

show the main function in the VSCode outline page

**Code sample and logs**

- Code sample

  all 2 files 

  `list.h` this is a open source doubly linked list header file.

  `main.cc` is my test code for `list.h`

  

  `list.h`

  ```c
  #ifndef _LINUX_LIST_H
  #define _LINUX_LIST_H
  
  /*
   * Simple doubly linked list implementation.
   *
   * Some of the internal functions ("__xxx") are useful when
   * manipulating whole lists rather than single entries, as
   * sometimes we already know the next/prev entries and we can
   * generate better code by using them directly rather than
   * using the generic single-entry routines.
   */
  
  struct list_node {
  	struct list_node *next, *prev;
  };
  
  #define LIST_HEAD_INIT(name) { &(name), &(name) }
  
  #define LIST_HEAD(name) \
  	struct list_node name = LIST_HEAD_INIT(name)
  
  static inline void INIT_LIST_HEAD(struct list_node *list)
  {
  	list->next = list;
  	list->prev = list;
  }
  
  /*
   * Insert a new entry between two known consecutive entries. 
   *
   * This is only for internal list manipulation where we know
   * the prev/next entries already!
   */
  static inline void __list_add(struct list_node *node,
  							  struct list_node *prev,
  							  struct list_node *next)
  {
  	next->prev = node;
  	node->next = next;
  	node->prev = prev;
  	prev->next = node;
  }
  
  /**
   * list_add - add a new entry
   * @new: new entry to be added
   * @head: list head to add it after
   *
   * Insert a new entry after the specified head.
   * This is good for implementing stacks.
   */
  static inline void list_add(struct list_node *node, struct list_node *head)
  {
  	__list_add(node, head, head->next);
  }
  
  /**
   * list_add_tail - add a new entry
   * @new: new entry to be added
   * @head: list head to add it before
   *
   * Insert a new entry before the specified head.
   * This is useful for implementing queues.
   */
  static inline void list_add_tail(struct list_node *node,
  								 struct list_node *head)
  {
  	__list_add(node, head->prev, head);
  }
  
  /*
   * Delete a list entry by making the prev/next entries
   * point to each other.
   *
   * This is only for internal list manipulation where we know
   * the prev/next entries already!
   */
  static inline void __list_del(struct list_node *prev, struct list_node *next)
  {
  	next->prev = prev;
  	prev->next = next;
  }
  
  /**
   * list_del - deletes entry from list.
   * @entry: the element to delete from the list.
   * Note: list_empty on entry does not return true after this, the entry is in an undefined state.
   */
  static inline void list_del(struct list_node *entry)
  {
  	__list_del(entry->prev, entry->next);
  }
  
  /**
   * list_move - delete from one list and add as another's head
   * @list: the entry to move
   * @head: the head that will precede our entry
   */
  static inline void list_move(struct list_node *list, struct list_node *head)
  {
  	__list_del(list->prev, list->next);
  	list_add(list, head);
  }
  
  /**
   * list_move_tail - delete from one list and add as another's tail
   * @list: the entry to move
   * @head: the head that will follow our entry
   */
  static inline void list_move_tail(struct list_node *list,
  								  struct list_node *head)
  {
  	__list_del(list->prev, list->next);
  	list_add_tail(list, head);
  }
  
  /**
   * list_empty - tests whether a list is empty
   * @head: the list to test.
   */
  static inline int list_empty(const struct list_node *head)
  {
  	return head->next == head;
  }
  
  static inline void __list_splice(struct list_node *list,
  								 struct list_node *head)
  {
  	struct list_node *first = list->next;
  	struct list_node *last = list->prev;
  	struct list_node *at = head->next;
  
  	first->prev = head;
  	head->next = first;
  
  	last->next = at;
  	at->prev = last;
  }
  
  /**
   * list_splice - join two lists
   * @list: the new list to add.
   * @head: the place to add it in the first list.
   */
  static inline void list_splice(struct list_node *list, struct list_node *head)
  {
  	if (!list_empty(list))
  		__list_splice(list, head);
  }
  
  /**
   * list_splice_init - join two lists and reinitialise the emptied list.
   * @list: the new list to add.
   * @head: the place to add it in the first list.
   *
   * The list at @list is reinitialised
   */
  static inline void list_splice_init(struct list_node *list,
  									struct list_node *head)
  {
  	if (!list_empty(list)) {
  		__list_splice(list, head);
  		INIT_LIST_HEAD(list);
  	}
  }
  
  /**
   * list_entry - get the struct for this entry
   * @ptr:	the &struct list_head pointer.
   * @type:	the type of the struct this is embedded in.
   * @member:	the name of the list_struct within the struct.
   */
  #define list_entry(ptr, type, member) \
  	((type *)((char *)(ptr)-(unsigned long)(&((type *)0)->member)))
  
  /**
   * list_for_each	-	iterate over a list
   * @pos:	the &struct list_head to use as a loop counter.
   * @head:	the head for your list.
   */
  #define list_for_each(pos, head) \
  	for (pos = (head)->next; pos != (head); pos = pos->next)
  
  /**
   * list_for_each_prev	-	iterate over a list backwards
   * @pos:	the &struct list_head to use as a loop counter.
   * @head:	the head for your list.
   */
  #define list_for_each_prev(pos, head) \
  	for (pos = (head)->prev; pos != (head); pos = pos->prev)
  
  /**
   * list_for_each_safe	-	iterate over a list safe against removal of list entry
   * @pos:	the &struct list_head to use as a loop counter.
   * @n:		another &struct list_head to use as temporary storage
   * @head:	the head for your list.
   */
  #define list_for_each_safe(pos, n, head) \
  	for (pos = (head)->next, n = pos->next; pos != (head); \
  		 pos = n, n = pos->next)
  
  /**
   * list_for_each_entry	-	iterate over list of given type
   * @pos:	the type * to use as a loop counter.
   * @head:	the head for your list.
   * @member:	the name of the list_struct within the struct.
   */
  #define list_for_each_entry(pos, head, member) \
  	for (pos = list_entry((head)->next, typeof (*pos), member); \
  		 &pos->member != (head); \
  		 pos = list_entry(pos->member.next, typeof (*pos), member))
  
  /**
   * Single-linked list. Added by Xie Han <xiehan@sogou-inc.com>.
   */
  
  struct slist_node {
  	struct slist_node *next;
  };
  
  struct slist_head {
  	struct slist_node first, *last;
  };
  
  #define SLIST_HEAD_INIT(name)	{ { (struct slist_node *)0 }, &(name).first }
  
  #define SLIST_HEAD(name) \
  	struct slist_head name = SLIST_HEAD_INIT(name)
  
  static inline void INIT_SLIST_HEAD(struct slist_head *list)
  {
  	list->first.next = (struct slist_node *)0;
  	list->last = &list->first;
  }
  
  static inline void slist_add_head(struct slist_node *node,
  								  struct slist_head *list)
  {
  	node->next = list->first.next;
  	list->first.next = node;
  	if (!node->next)
  		list->last = node;
  }
  
  static inline void slist_add_tail(struct slist_node *node,
  								  struct slist_head *list)
  {
  	node->next = (struct slist_node *)0;
  	list->last->next = node;
  	list->last = node;
  }
  
  static inline void slist_add_after(struct slist_node *node,
  								   struct slist_node *prev,
  								   struct slist_head *list)
  {
  	node->next = prev->next;
  	prev->next = node;
  	if (!node->next)
  		list->last = node;
  }
  
  static inline void slist_del_head(struct slist_head *list)
  {
  	list->first.next = list->first.next->next;
  	if (!list->first.next)
  		list->last = &list->first;
  }
  
  static inline void slist_del_after(struct slist_node *prev,
  								   struct slist_head *list)
  {
  	prev->next = prev->next->next;
  	if (!prev->next)
  		list->last = prev;
  }
  
  static inline int slist_empty(struct slist_head *list)
  {
  	return !list->first.next;
  }
  
  static inline void __slist_splice(struct slist_head *list,
  								  struct slist_node *at,
  								  struct slist_head *head)
  {
  	list->last->next = at->next;
  	at->next = list->first.next;
  	if (!list->last->next)
  		head->last = list->last;
  }
  
  static inline void slist_splice(struct slist_head *list,
  								struct slist_node *at,
  								struct slist_head *head)
  {
  	if (!slist_empty(list))
  		__slist_splice(list, at, head);
  }
  
  static inline void slist_splice_init(struct slist_head *list,
  									 struct slist_node *at,
  									 struct slist_head *head)
  {
  	if (!slist_empty(list)) {
  		__slist_splice(list, at, head);
  		INIT_SLIST_HEAD(list);
  	}
  }
  
  #define slist_entry(ptr, type, member) \
  	((type *)((char *)(ptr)-(unsigned long)(&((type *)0)->member)))
  
  #define slist_for_each(pos, head) \
  	for (pos = (head)->first.next; pos; pos = pos->next)
  
  #define slist_for_each_safe(pos, prev, head) \
  	for (prev = &(head)->first, pos = prev->next; pos; \
  		 prev = prev->next == pos ? pos : prev, pos = prev->next)
  
  #define slist_for_each_entry(pos, head, member) \
  	for (pos = slist_entry((head)->first.next, typeof (*pos), member); \
  		 &pos->member != (struct slist_node *)0; \
  		 pos = slist_entry(pos->member.next, typeof (*pos), member))
  
  #endif
  
  ```

  

  `main.cc` 

  ```c
  #include "list.h"
  #include <stdio.h>
  #include <stdlib.h>
  
  #define MAX_CNT 10
  
  struct xy_data
  {
      int x;
      int y;
  };
  
  struct data_entry
  {
      struct xy_data data;
      struct list_node list;
  };
  
  int main()
  {
      struct list_node data_queue;
      struct data_entry *p_new_data;
  
      INIT_LIST_HEAD(&data_queue);
  
      // 1. 插入数据
      for (int i = 0; i < MAX_CNT; i++)
      {
          p_new_data = (struct data_entry *)malloc(sizeof(struct data_entry));
          if (0 == p_new_data)
              return -1;
  
          p_new_data->data.x = i;
          p_new_data->data.y = i;
  
          list_add_tail(&p_new_data->list, &data_queue);
      }
  
      // 2. 展示数据
      struct list_node *pos, *tmp;
      struct data_entry *entry;
  
      list_for_each_safe(pos, tmp, &data_queue)
      {
          // zwd: list_entry 这个宏的写法很魔幻，对空指针的结构体成员取地址，可以得到该成员的偏移量，学废了学废了。
          entry = list_entry(pos, struct data_entry, list);
  
          printf("%d-%d\n", entry->data.x, entry->data.y);
      }
  
      // 3. 销毁链表
      list_for_each_safe(pos, tmp, &data_queue)
      {
          entry = list_entry(pos, struct data_entry, list);
          list_del(pos);
          free(entry);
      }
  
      return 0;
  }
  ```

  

- Configurations in `c_cpp_properties.json`

- Logs from running `C/C++: Log Diagnostics` from the VS Code command palette

- Logs from [the language server logging](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server)

**Screenshots**

this is outline screenshot for main.cc
![Snipaste_2022-05-30_17-11-48](https://user-images.githubusercontent.com/37608409/170959181-415b1014-e9f9-486a-9283-6f47506e567e.png)

>

**Additional context**

**Comment by sean-mcmanus** [comment]
The issue doesn't repro with 1.10.3 (pre-release), although prior to that we would have considered the bug "by design' and it would have been expected that the macro would need to be added to a cpp.hint file. Let us know if you're still seeing a bug with that version.

**Comment by wenda84** [comment]
yes，it is fixed in v1.10.3 (pre-release). thx.

![Snipaste_2022-06-01_09-17-52](https://user-images.githubusercontent.com/37608409/171308958-009227e9-3926-4988-8d92-47a6dd957344.png)
